GitDiffMargin/Git/GitCommands.cs

```diff
@@ -282,5 +325,90 @@ static bool HasPreamble(string file, Encoding encoding)
 
             return true;
         }
+
+        private string AdjustPath(string fullPath)
+        {
+            // Right now the only extension is for *.csproj and *.vbproj, which are opened from a temporary location
+            switch (Path.GetExtension(fullPath).ToLowerInvariant())
+            {
+            case ".csproj":
+            case ".vbproj":
+                break;
+
+            default:
+                return fullPath;
+            }
+
+            // CPS will open the file in %TEMP%\{random name}\{ProjectFileName}
+            string directoryName = Path.GetDirectoryName(fullPath);
+            if (string.IsNullOrEmpty(directoryName))
+                return fullPath;
+
+            directoryName = Path.GetDirectoryName(directoryName);
+            if (!Path.GetTempPath().Equals(directoryName + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase))
+                return fullPath;
+
+            IVsSolution solution = _serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            if (solution == null)
+                return fullPath;
+
+            if (!ErrorHandler.Succeeded(solution.GetProjectEnum((uint)__VSENUMPROJFLAGS.EPF_LOADEDINSOLUTION, Guid.Empty, out IEnumHierarchies ppenum))
+                || ppenum == null)
+            {
+                return fullPath;
+            }
+
+            List<string> projectFiles = new List<string>();
+            IVsHierarchy[] hierarchies = new IVsHierarchy[1];
+            while (true)
+            {
+                int hr = ppenum.Next((uint)hierarchies.Length, hierarchies, out uint fetched);
+                if (!ErrorHandler.Succeeded(hr))
+                    return fullPath;
+
+                for (uint i = 0; i < fetched; i++)
+                {
+                    if (!(hierarchies[0] is IVsProject project))
+                        continue;
+
+                    if (!ErrorHandler.Succeeded(project.GetMkDocument((uint)VSConstants.VSITEMID.Root, out string projectFilePath)))
+                        continue;
+
+                    if (!Path.GetFileName(projectFilePath).Equals(Path.GetFileName(fullPath), StringComparison.OrdinalIgnoreCase))
+                        continue;
+
+                    projectFiles.Add(projectFilePath);

```

Comment:
@sharwell, you can use https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.ivshierarchy.getproperty.aspx, with `__VSHPROPID5.VSHPROPID_ProjectCapabilities`. https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.shell.interop.__vshpropid5.aspx